pkg/workloads/containerd/watcher.go

```diff
@@ -52,82 +51,72 @@ const (
 	eventQueueBufferSize = 100
 )
 
-// containerEvents holds per-container queues for events
-type containerEvents struct {
-	sync.Mutex
-	events map[string]chan dTypesEvents.Message
-}
-
-func (ce *containerEvents) enqueueByContainerID(e dTypesEvents.Message) {
-	ce.Lock()
-	defer ce.Unlock()
-
-	if _, found := ce.events[e.Actor.ID]; !found {
-		q := make(chan dTypesEvents.Message, eventQueueBufferSize)
-		ce.events[e.Actor.ID] = q
-		go processContainerEvents(e.Actor.ID, q)
-	}
-	ce.events[e.Actor.ID] <- e
-}
-
-func (ce *containerEvents) reapEmpty() {
-	ce.Lock()
-	defer ce.Unlock()
-
-	for id, q := range ce.events {
-		if len(q) == 0 {
-			close(q)
-			delete(ce.events, id)
-		}
-	}
-}
-
 func shortContainerID(id string) string {
 	return id[:10]
 }
 
 // EnableEventListener watches for docker events. Performs the plumbing for the
 // containers started or dead.
 func EnableEventListener() error {
-	eventQueue := containerEvents{events: make(map[string]chan dTypesEvents.Message)}
 	since := time.Now()
-	syncWithRuntime()
-
-	eo := dTypes.EventsOptions{Since: strconv.FormatInt(since.Unix(), 10)}
-	r, err := dockerClient.Events(ctx.Background(), eo)
-	if err != nil {
+	ws := newWatcherState()
+	syncWithRuntime(ws, since)
+	if err := startEventListener(ws, since); err != nil {
 		return err
 	}
 
-	go listenForDockerEvents(&eventQueue, r)
-
 	// start a go routine which periodically synchronizes containers
 	// managed by the local container runtime and checks if any of them
 	// need to be managed by Cilium. This is a fall back mechanism in case
 	// an event notification has been lost.
 	go func() {
+		currentWatcherState := ws // inherit this from EnableEventListener's scope
 		for {
 			time.Sleep(syncRateDocker)
-			syncWithRuntime()
+			since := time.Now()
+			newWS := newWatcherState()
```

Comment:
Do we need to start a new watcher on every loop? I think this can be moved to line 79/80

Reply:
We do because syncWithRuntime populates the state. Instead of iterating over all the containers once to determine if we're out of sync, then doing it again to create the new state, I do it once. It's not very clean but I didn't want to change the current logic even more than I have.